src/sc2api/sc2_client.cc

```diff
@@ -2035,73 +2039,60 @@ void ControlImp::IssueUnitDestroyedEvents() {
-    if (found == observation_imp_->units_previous_map_.end()) {
-        client_.OnUnitIdle(unit);
-        return;
+    for (auto unit : observation_imp_->unit_pool_.GetUnitsEnteringVision())
+    {
+        if (unit->alliance == Unit::Alliance::Enemy && unit->display_type == Unit::DisplayType::Visible)
+            client_.OnUnitEnterVision(unit);
     }
+}
 
-    // Otherwise get that unit from the previous list and verify it's state changed to idle.
-    const Unit& unit_previous = found->second;
-
-    if (!unit_previous.orders.empty()) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueUnitDamagedEvents() {
+    for (auto const *u : observation_imp_->unit_pool_.GetDamagedUnits())
+        client_.OnUnitDamaged(u);
+}
 
-    // If the unit had less than 1.0 build progress in the last stop this is the first time it's active.
-    if (unit_previous.build_progress < 1.0f) {
-        client_.OnUnitIdle(unit);
-        return;
-    }
+void ControlImp::IssueIdleEvents(const std::vector<Tag>& commands) {
 
-    // Iterate the issued commands, if a unit exists in that list but does not currently have orders
-    // the order must have failed. Reissue the OnUnitIdle event in that case.
+    auto& unit_pool = observation_imp_->unit_pool_;
+    // identify idled units where commands were issued last step, but units have no orders now (maybe failed, maybe executed instantly)
     for (auto t : commands) {
-        if (t == unit->tag) {
-            client_.OnUnitIdle(unit);
-            break;
-        }
+        const auto* unit = unit_pool.GetExistingUnit(t);
+        if (unit && unit->orders.empty())
+            unit_pool.AddUnitIdled(unit);
     }
-}
 
-void ControlImp::IssueBuildingCompletedEvent(const Unit* unit) {
-    // If the units build progress is complete but it previously wasn't call construction complete
-    if (!unit || unit->build_progress < 1.0f) {
-        return;
-    }
+    // add newly created units (if they are completed)
+    for (auto const* u : unit_pool.GetNewUnits())
+        if (u->build_progress == 1.0f && u->orders.empty())
+            unit_pool.AddUnitIdled(u);
 
-   auto found = observation_imp_->units_previous_map_.find(unit->tag);
+    // add completed buildings (check for orders, to avoid possibility of classifying a newly constructed stargate that is
+    // transforming to warpgate as idle -- not sure that happens though)
+    for (auto const* u : unit_pool.GetCompletedBuildings())
+        if (u->build_progress == 1.0f && u->orders.empty())
+            unit_pool.AddUnitIdled(u);
 
-    if (found == observation_imp_->units_previous_map_.end()) {
-        return;
-    }
+    // send only one idle event for any unit in any frame
+    auto const& idled_units = unit_pool.GetIdledUnits();
+    std::for_each(idled_units.begin(), idled_units.end(), [this](auto const * u) {client_.OnUnitIdle(u); });
```

Comment:
Style: this is a bit inconsistent now as the code above uses simple `for` and here we use `for_each`.

Reply:
your right. I think because my other lambdas weren't compiling for me. I'll use the simple for block for consistency.

Reply:
I personally prefer `for_each` although it could be harder to implement and read. So, it would be cool to switch to `for_each` whenever possible.

Reply:
I agree it is harder to read and write. Also, for_each is stylistically inconsistent with the current sc2api (it is not used anywhere). Maybe we can wait on this?

Reply:
Ok, no problem.